adam-core/src/main/scala/org/bdgenomics/adam/serialization/BinaryDecoderWithLRUCache.scala

```diff
@@ -0,0 +1,120 @@
+/**
+ * Licensed to Big Data Genomics (BDG) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The BDG licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bdgenomics.adam.serialization
+
+import java.io.{ IOException, InputStream }
+import java.nio.ByteBuffer
+import java.util
+import java.util.Map.Entry
+
+import org.apache.avro.io.{ BinaryDecoder, Decoder, DecoderFactory }
+import org.apache.avro.util.Utf8
+
+/**
+ * Note: A LinkedHashMap takes 40 * SIZE + 4 * CAPACITY bytes. If the size == capacity, then the memory
+ * footprint is CAPACITY * 44 bytes.
+ */
+private[serialization] class LRUCache[K, V](maxCapacity: Int, loadFactor: Double = 0.75f)
+    extends util.LinkedHashMap[K, V](maxCapacity, 0.75f, true) {
+  assert(maxCapacity > 0, "LRU cache maxCapacity should be greater than 0")
+  assert(0 < loadFactor && loadFactor <= 1.0, "LRU load factor must be (0,1.0]")
+  override def removeEldestEntry(eldest: Entry[K, V]): Boolean = {
+    size() > maxCapacity
+  }
+}
+
+/**
+ * An Avro binary decoder that keep a LRU cache of strings to minimize memory pressure.
+ *
+ * WARNING: This class is not thread-safe! This isn't an issue since Spark guarantees that only
+ * a single thread will access it.
+ *
+ * A amount of memory strings take in Java is 8 * (int) ((((#chars) * 2) + 45) / 8) bytes, e.g. a string that is
+ * 128 characters long will take, 8 * ((((128) * 2) + 45) / 8) = 296 bytes. This BinaryDecoder is not for general
+ * use but is optimized for ADAM -omics data, where strings are typically ~128 characters long.
```

Comment:
So you're recommending a string length cutoff of say 64 bytes so that referenceName, readName, cigar, etc strings are cached but sequence and quality are likely not? I like that idea.